tracing/src/main/java/com/palantir/tracing/Trace.java

```diff
@@ -282,13 +307,19 @@ boolean isObservable() {
             return false;
         }
 
+        @Override
+        Optional<String> getTopSpanId() {
+            return topSpanId;
+        }
+
         @Override
         Optional<String> getOriginatingSpanId() {
             return originatingSpanId;
         }
 
         @Override
         Trace deepCopy() {
+            // TODO: shouldn't this preserve originatingSpanId / topSpanId?
             return new Unsampled(numberOfSpans, getTraceId());
```

Comment:
Copying an `Unsampled` trace discards originating / top span ids. (a sampled one does not). This seems like a bug?

Reply:
Chatted with @dansanduleac - we think this means requests made in deferred traces won't have an originating span id header set

Reply:
Actually looked at the code and there is something broken here indeed, but not where you think.

DeferredTrace actually copies twice: the first time in its constructor, it's only as a hack, in order to get access to the Trace, but the 2nd time it's to save (and then restore) the original trace.
It's this 2nd time where interesting things can happen.

Note that DeferredTrace keeps track of the `parentSpanId` and the inner "deferred" Trace it creates is actually empty, as in you don't carry around the existing span stack. When you call `.withTrace` to use the deferred trace, it then calls `fastStartSpan` with that saved `parentSpanId`, thus ensuring that the new Trace initializes its `originatingSpanId` correctly from that.
All good so far.

Now it gets tricky when restoring the original trace.
We create a new Trace, losing the information as you correctly point out.
For Sampled, *new* spans are able to recover the correct `originatingSpanId` because `Sampled.getOriginatingSpanId()` grabs it from the last item in the stack (i.e. the top).
For *Unsampled* though, `getOriginatingSpanId` will now be cleared and unable to recover the original value. 🌶🌶🌶

Reply:
>Note that DeferredTrace keeps track of the parentSpanId and the inner "deferred" Trace it creates is actually empty, as in you don't carry around the existing span stack.

Oh so curiously the originating span id of a deferred trace is the innermost span that created the deferred trace, not the originating span id of the trace at that point?

Not sure if this is intentional.

I guess we don't copy the entire stack because we expect those spans to get closed before the deferred trace actually gets to run?

Reply:
I think you are right... that doesn't sound right to me either.

I think we don't copy the entire stack because we want to enforce those spans to not be involved in the deferred span trace, i.e. you are not allowed to close them.